File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    
    for i in range(len(computers)):
        if dfs(i, computers):
            answer += 1
    
    return answer

def dfs(v, computers):
    if computers[v][v] == 0:
        return False
    
    computers[v][v] = 0
    for i in range(len(computers)):
        if computers[v][i] == 1:
            computers[v][i] = 0
            dfs(i, computers)
    return True
    

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    int[] parent;
    
    public int solution(int n, int[][] computers) {
        parent = new int[n];
        
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i <= j || computers[i][j] == 0) continue;

                union(i, j);
            }
        }
        
        Set<Integer> set = new HashSet<>();
        
        for (int i = 0; i < n; i++) {
            set.add(findParent(i));
        }
        
        return set.size();
    }
    
    int findParent(int x) {
        if (parent[x] == x) {
            return x;
        }
        
        return parent[x] = findParent(parent[x]);
    }
    
    void union(int a, int b) {
        int rootA = findParent(a);
        int rootB = findParent(b);
        
        if (rootA != rootB) {
            parent[rootB] = rootA;
        }
    }
}